File: src/utils/matlab/libs/plotTracking.m
function plotTracking(channelList, trackResults, settings)
% This function plots the tracking results for the given channel list.
%
% plotTracking(channelList, trackResults, settings)
%
%   Inputs:
%       channelList     - list of channels to be plotted.
%       trackResults    - tracking results from the tracking function.
%       settings        - receiver settings.

%--------------------------------------------------------------------------
%                           SoftGNSS v3.0
%
% Copyright (C) Darius Plausinaitis
% Written by Darius Plausinaitis
%--------------------------------------------------------------------------
%This program is free software; you can redistribute it and/or
%modify it under the terms of the GNU General Public License
%as published by the Free Software Foundation; either version 2
%of the License, or (at your option) any later version.
%
%This program is distributed in the hope that it will be useful,
%but WITHOUT ANY WARRANTY; without even the implied warranty of
%MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%GNU General Public License for more details.
%
%You should have received a copy of the GNU General Public License
%along with this program; if not, write to the Free Software
%Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
%USA.
%--------------------------------------------------------------------------


% Protection - if the list contains incorrect channel numbers
channelList = intersect(channelList, 1:settings.numberOfChannels);

%=== For all listed channels ==============================================
for channelNr = channelList
    
    %% Select (or create) and clear the figure ================================
    % The number 200 is added just for more convenient handling of the open
    % figure windows, when many figures are closed and reopened.
    % Figures drawn or opened by the user, will not be "overwritten" by
    % this function.
    
    figure(channelNr +200);
    clf(channelNr +200);
    set(channelNr +200, 'Name', ['Channel ', num2str(channelNr), ...
        ' (PRN ', ...
        num2str(trackResults(channelNr).PRN(end-1)), ...
        ') results']);
    
    %% Draw axes ==============================================================
    % Row 1
    handles(1, 1) = subplot(4, 3, 1);
    handles(1, 2) = subplot(4, 3, [2 3]);
    % Row 2
    handles(2, 1) = subplot(4, 3, 4);
    handles(2, 2) = subplot(4, 3, [5 6]);
    % Row 3
    handles(3, 1) = subplot(4, 3, 7);
    handles(3, 2) = subplot(4, 3, 8);
    handles(3, 3) = subplot(4, 3, 9);
    % Row 4
    handles(4, 1) = subplot(4, 3, 10);
    handles(4, 2) = subplot(4, 3, 11);
    handles(4, 3) = subplot(4, 3, 12);
    
    
    %% Plot all figures =======================================================
    
    timeAxisInSeconds = (1:settings.msToProcess)/1000;
    
    %----- Discrete-Time Scatter Plot ---------------------------------
    plot(handles(1, 1), trackResults(channelNr).I_P,...
        trackResults(channelNr).Q_P, ...
        '.');
    
    grid  (handles(1, 1));
    axis  (handles(1, 1), 'equal');
    title (handles(1, 1), 'Discrete-Time Scatter Plot');
    xlabel(handles(1, 1), 'I prompt');
    ylabel(handles(1, 1), 'Q prompt');
    
    %----- Nav bits ---------------------------------------------------
    plot  (handles(1, 2), timeAxisInSeconds, ...
        trackResults(channelNr).I_P);
    
    grid  (handles(1, 2));
    title (handles(1, 2), 'Bits of the navigation message');
    xlabel(handles(1, 2), 'Time (s)');
    axis  (handles(1, 2), 'tight');
    
    %----- PLL discriminator unfiltered--------------------------------
    plot  (handles(2, 1), timeAxisInSeconds, ...
        trackResults(channelNr).pllDiscr, 'r');
    
    grid  (handles(2, 1));
    axis  (handles(2, 1), 'tight');
    xlabel(handles(2, 1), 'Time (s)');
    ylabel(handles(2, 1), 'Amplitude');
    title (handles(2, 1), 'Raw PLL discriminator');
    
    %----- Correlation ------------------------------------------------
    plot(handles(2, 2), timeAxisInSeconds, ...
        [sqrt(trackResults(channelNr).I_E.^2 + ...
        trackResults(channelNr).Q_E.^2)', ...
        sqrt(trackResults(channelNr).I_P.^2 + ...
        trackResults(channelNr).Q_P.^2)', ...
        sqrt(trackResults(channelNr).I_L.^2 + ...
        trackResults(channelNr).Q_L.^2)'], ...
        '-*');
    
    grid  (handles(2, 2));
    title (handles(2, 2), 'Correlation results');
    xlabel(handles(2, 2), 'Time (s)');
    axis  (handles(2, 2), 'tight');
    
    hLegend = legend(handles(2, 2), '$\sqrt{I_{E}^2 + Q_{E}^2}$', ...
        '$\sqrt{I_{P}^2 + Q_{P}^2}$', ...
        '$\sqrt{I_{L}^2 + Q_{L}^2}$');
    
    %set interpreter from tex to latex. This will draw \sqrt correctly
    set(hLegend, 'Interpreter', 'Latex');
    
    %----- PLL discriminator filtered----------------------------------
    plot  (handles(3, 1), timeAxisInSeconds, ...
        trackResults(channelNr).pllDiscrFilt(1:settings.msToProcess), 'b');
    
    grid  (handles(3, 1));
    axis  (handles(3, 1), 'tight');
    xlabel(handles(3, 1), 'Time (s)');
    ylabel(handles(3, 1), 'Amplitude');
    title (handles(3, 1), 'Filtered PLL discriminator');
    
    %----- DLL discriminator unfiltered--------------------------------
    plot  (handles(3, 2), timeAxisInSeconds, ...
        trackResults(channelNr).dllDiscr, 'r');
    
    grid  (handles(3, 2));
    axis  (handles(3, 2), 'tight');
    xlabel(handles(3, 2), 'Time (s)');
    ylabel(handles(3, 2), 'Amplitude');
    title (handles(3, 2), 'Raw DLL discriminator');
    
    %----- DLL discriminator filtered----------------------------------
    plot  (handles(3, 3), timeAxisInSeconds, ...
        trackResults(channelNr).dllDiscrFilt, 'b');
    
    grid  (handles(3, 3));
    axis  (handles(3, 3), 'tight');
    xlabel(handles(3, 3), 'Time (s)');
    ylabel(handles(3, 3), 'Amplitude');
    title (handles(3, 3), 'Filtered DLL discriminator');
    
    %----- CNo for signal----------------------------------
    plot  (handles(4, 1), timeAxisInSeconds, ...
        trackResults(channelNr).CNo(1:settings.msToProcess), 'b');
    
    grid  (handles(4, 1));
    axis  (handles(4, 1), 'tight');
    xlabel(handles(4, 1), 'Time (s)');
    ylabel(handles(4, 1), 'CNo (dB-Hz)');
    title (handles(4, 1), 'Carrier to Noise Ratio');
    
    %----- Carrier Frequency --------------------------------
    plot  (handles(4, 2), timeAxisInSeconds(2:end), ...
        trackResults(channelNr).carrFreq(2:settings.msToProcess), 'Color',[0.42 0.25 0.39]);
    
    grid  (handles(4, 2));
    axis  (handles(4, 2));
    xlabel(handles(4, 2), 'Time (s)');
    ylabel(handles(4, 2), 'Freq (hz)');
    title (handles(4, 2), 'Carrier Freq');
    
    %----- Code Frequency----------------------------------
    %--- Skip sample 0 to help with results display
    plot  (handles(4, 3), timeAxisInSeconds(2:end), ...
        trackResults(channelNr).codeFreq(2:settings.msToProcess), 'Color',[0.2 0.3 0.49]);
    
    grid  (handles(4, 3));
    axis  (handles(4, 3), 'tight');
    xlabel(handles(4, 3), 'Time (s)');
    ylabel(handles(4, 3), 'Freq (Hz)');
    title (handles(4, 3), 'Code Freq');
    
end % for channelNr = channelList

File: utils/python/lib/plotTracking.py
"""
 plotTracking.py

 This function plots the tracking results for the given channel list.

 Irene Pérez Riega, 2023. iperrie@inta.es

 plotTracking(channelList, trackResults, settings)

   Args:
       channelList     - list of channels to be plotted.
       trackResults    - tracking results from the tracking function.
       settings        - receiver settings.

   Modifiable in the file:
       fig_path        - Path where plots will be save

 -----------------------------------------------------------------------------

 GNSS-SDR is a Global Navigation Satellite System software-defined receiver.
 This file is part of GNSS-SDR.

 Copyright (C) 2022  (see AUTHORS file for a list of contributors)
 SPDX-License-Identifier: GPL-3.0-or-later

 -----------------------------------------------------------------------------
"""

import numpy as np
import os
import matplotlib.pyplot as plt

def plotTracking(channelNr, trackResults, settings):

    # ---------- CHANGE HERE:
    fig_path = '/home/labnav/Desktop/TEST_IRENE/PLOTS/PlotTracking'

    if not os.path.exists(fig_path):
        os.makedirs(fig_path)

    # Protection - if the list contains incorrect channel numbers
    if channelNr in list(range(1,settings["numberOfChannels"]+1)):

        plt.figure(figsize=(1920 / 120, 1080 / 120))
        plt.clf()
        plt.gcf().canvas.set_window_title(
            f'Channel {channelNr} (PRN '
            f'{trackResults[channelNr-1]["PRN"][0]}) results')
        plt.subplots_adjust(left=0.1, right=0.9, top=0.9, bottom=0.1,
                            hspace=0.4, wspace=0.4)
        plt.tight_layout()

        # Extract timeAxis and time_label
        if 'prn_start_time_s' in trackResults[channelNr-1]:
            timeAxis = trackResults[channelNr-1]['prn_start_time_s']
            time_label = 'RX Time (s)'
        else:
            timeAxis = np.arange(1, len(trackResults[channelNr-1]['PRN']) + 1)
            time_label = 'Epoch'


        # Row 1 ==============================================================
        # Discrete-Time Scatter Plot
        plt.subplot(4, 3, 1)
        plt.plot(trackResults[channelNr-1]['I_P'],
                 trackResults[channelNr-1]['Q_P'], marker='.', markersize=1,
                 linestyle=' ')
        plt.grid()
        plt.axis('equal')
        plt.title('Discrete-Time Scatter Plot', fontweight='bold')
        plt.xlabel('I prompt')
        plt.ylabel('Q prompt')

        # Nav bits
        plt.subplot(4, 3, (2, 3))
        plt.plot(timeAxis, trackResults[channelNr-1]['I_P'], linewidth=1)
        plt.grid()
        plt.title('Bits of the navigation message', fontweight='bold')
        plt.xlabel(time_label)
        plt.axis('tight')

        # Row 2 ==============================================================
        # Raw PLL discriminator unfiltered
        plt.subplot(4, 3, 4)
        plt.plot(timeAxis, trackResults[channelNr-1]['pllDiscr'],
                 color='r', linewidth=1)
        plt.grid()
        plt.axis('tight')
        plt.xlabel(time_label)
        plt.ylabel('Amplitude')
        plt.title('Raw PLL discriminator', fontweight='bold')

        # Correlation results
        plt.subplot(4, 3, (5, 6))
        corr_data = [
            np.sqrt(trackResults[channelNr-1]['I_E'] ** 2 +
                    trackResults[channelNr-1]['Q_E'] ** 2),
            np.sqrt(trackResults[channelNr-1]['I_P'] ** 2 +
                    trackResults[channelNr-1]['Q_P'] ** 2),
            np.sqrt(trackResults[channelNr-1]['I_L'] ** 2 +
                    trackResults[channelNr - 1]['Q_L'] ** 2)
        ]

        line = []
        colors = ['b', '#FF6600', '#FFD700', 'purple', 'g']

        for i, data in enumerate(corr_data):
            line.append(plt.plot(timeAxis, data,
                                 label=f'Data {i+1}', color=colors[i],
                                 marker='*', linestyle=' ', linewidth=1))

        plt.grid()
        plt.title('Correlation results', fontweight='bold')
        plt.xlabel(time_label)
        plt.axis('tight')
        plt.legend([r'$\sqrt{I_{VE}^2 + Q_{VE}^2}$',
                    r'$\sqrt{I_{E}^2 + Q_{E}^2}$',
                    r'$\sqrt{I_{P}^2 + Q_{P}^2}$',
                    r'$\sqrt{I_{L}^2 + Q_{L}^2}$',
                    r'$\sqrt{I_{VL}^2 + Q_{VL}^2}$'], loc='best')

        # Row 3 ==============================================================
        # Filtered PLL discriminator
        plt.subplot(4, 3, 7)
        plt.plot(timeAxis, trackResults[channelNr-1]['pllDiscrFilt'],
                 'b', linewidth=1)
        plt.grid()
        plt.axis('tight')
        plt.xlabel(time_label)
        plt.ylabel('Amplitude')
        plt.title('Filtered PLL discriminator', fontweight='bold')

        # Raw DLL discriminator unfiltered
        plt.subplot(4, 3, 8)
        plt.plot(timeAxis, trackResults[channelNr-1]['dllDiscr'],
                 'r', linewidth=1)
        plt.grid()
        plt.axis('tight')
        plt.xlabel(time_label)
        plt.ylabel('Amplitude')
        plt.title('Raw DLL discriminator', fontweight='bold')

        # Filtered DLL discriminator
        plt.subplot(4, 3, 9)
        plt.plot(timeAxis, trackResults[channelNr-1]['dllDiscrFilt'],
                 'b', linewidth=1)
        plt.grid()
        plt.axis('tight')
        plt.xlabel(time_label)
        plt.ylabel('Amplitude')
        plt.title('Filtered DLL discriminator', fontweight='bold')

        # Row 4 ==============================================================
        # CNo for signal
        plt.subplot(4, 3, 10)
        plt.plot(timeAxis, trackResults[channelNr-1]['CNo'], 'b',
                 linewidth=1)
        plt.grid()
        plt.axis('equal')
        plt.xlabel('Time (s)')
        plt.ylabel('CNo (dB-Hz)')
        plt.title('Carrier to Noise Ratio', fontweight='bold')

        # Carrier Frequency
        plt.subplot(4, 3, 11)
        plt.plot(timeAxis, trackResults[channelNr-1]['carrFreq'],
                 marker='.', markersize=1, linestyle=' ')
        plt.grid()
        plt.axis('equal')
        plt.xlabel('Time (s)')
        plt.ylabel('Freq (hz)')
        plt.title('Carrier Frequency', fontweight='bold')

        # Code Frequency
        # Skip sample 0 to help with results display
        plt.subplot(4, 3, 12)
        plt.plot(timeAxis, trackResults[channelNr-1]['codeFreq'],
                 marker='.', markersize=1, linestyle=' ')
        plt.grid()
        plt.axis('equal')
        plt.xlabel('Time (s)')
        plt.ylabel('Freq (Hz)')
        plt.title('Code Frequency',fontweight='bold')

        plt.tight_layout()
        plt.savefig(os.path.join(fig_path,
                                 f'trk_dump_ch{channelNr}_PRN_'
                                 f'{trackResults[channelNr - 1]["PRN"][-1]}'
                                 f'.png'))
        plt.show()
